### ISSUE TITLE
federation: non-federation DNS lookup can return federation CNAME

### ISSUE THREAD

**Issue by nikhiljindal** [kubernetes/kubernetes#27969]
With the current KubeDNS code, it can happen that a pod requests `mysvc.somens` and resolve.conf adds `myns.svc.cluster.local` as a search path so that KubeDNS gets `mysvc.somns.myns.svc.cluster.local`.
If there is a federation with the name `myns`, isFederationQuery in KubeDNS will think this is a federation query and try to resolve `mysvc.somens.svc.cluster.local`. If there is a local service resolving to that DNS, then user will get that as expected (all is fine). But if there isnt, then KubeDNS will return `mysvc.somens.myns.svc.myzone.myregion.mydomain`. If there is no federation service with that name and in that namespace, then user will still get an NXDOMAIN as expected. But if there is, then user might be pointed to a service in another cluster, which user may or may not have wanted.

We can consider this a feature, if users want it or should fix it, if we consider it a bug.

Filing this to keep track.

@kubernetes/sig-cluster-federation

**Comment by nikhiljindal** [comment]
For the record, @quinton-hoole and I had discussed a potential solution which we decided not to do.

The idea was to update kubedns.isFederationQuery() to look for the pattern `svcname.nsname.fedname.nsname.svc.domain` instead of looking for `svcname.nsname.fedname.svc.domain`. That will fix the problem, but it doesnt seem logically and syntactically correct to resolve `svcname.nsname.fedname.nsname.svc.domain` instead of resolving `svcname.nsname.fedname.svc.domain`. And this solution works only because of the current ordering of search paths in resolve.conf (we first append `nsname.svc.domain` and then `svc.domain`). It can break if that order changes.
